### ISSUE TITLE
 Bio::Tools::GFF doesn't write eukaryotic multi-exon genes correctly

### ISSUE THREAD

**Issue by krobison13** [bioperl/bioperl-live#369]
When writing GFF, the same frame is assigned to every range in a multi-exon gene rather than correctly assigning 0,1 or 2 to specify the frame

[ Twitter note including image of the offending loop](https://twitter.com/OmicsOmicsBlog/status/1512126755196256264?s=20&t=47TNCWLA8pnXiIzyGut47g)

**Comment by hyphaltip** [comment]
thanks for reporting a bug rather than just twitter rant!

maybe others can look at this too @cjfields --  its been 15+ years with that code - I think this is more about assumptions about features vs locations here that are obscuring the problem you describe.

If one is reading and writing multi-exons as individual features which is the typical way the frame is encoded this all works as planned - but if a single feature is encoded as a split-location  - it isn't encoded in a multi-location genbank file. If you wanted it to be computed from the data that might be helpful but it also make assumptions about the Generic feature being a CDS. This goes back to pre-GFF3 when the assumptions about how parent/child relationships were encoded and there were multiple interpretations of how to do this from gff1->gff2->gff2.5 /gtf etc.

maybe you can show input data that you used - are you are converting genbank to GFF and expecting frame to be computed and the assumption that it is a CDS with a frame to be carried through?

**Comment by krobison13** [comment]
The zip file has a simple Genbank-formatted entry and a simple program that exposes the problem -- the  correct sequence of frames is 0,0,2,1,0

[gff-bug-reveal.zip](https://github.com/bioperl/bioperl-live/files/8452913/gff-bug-reveal.zip)

**Comment by cjfields** [comment]
Yeah I agree w/ @hyphaltip , I suspect there's bit rot from prior logical assumptions that have changed over time.  I also vaguely recall Bio::Tools::GFF was to be deprecated in preference to [Bio::DB::SeqFeature](https://metacpan.org/pod/Bio::DB::SeqFeature), though I'm not sure that is still the case. 

Would it be worth looking into Bio::Tools::GFF or should we check Bio::DB::SeqFeature?  If @scottcain around, maybe he would know?  I think there was a GenBank-to-GFF conversion script for Bio::DB::SeqFeature (maybe within the GBrowse2 code?), we could check to see if if gives the correct frames.

**Comment by scottcain** [comment]
Ugh. Bio::Tools::GFF was old and janky a long time ago and should probably be marked as such, since I don't think it is likely to have improved with age. It is hard to remember the logic that went into that bit of code (I don't recall if I wrote it--I hope not--but I certainly might have!). I think @cjfields is right about there having been a GB to GFF3 script, but I don't recall where it lived.

There is a script with GBrowse, https://github.com/GMOD/GBrowse/blob/master/bin/load_genbank.pl, but it loads into a Bio::DB::GFF database (so, GFF2 and mysql or postgres). I don't have the time to do the code archeology to determine if it handles strand better.

**Comment by carandraug** [comment]
> I think @cjfields is right about there having been a GB to GFF3 script, but I don't recall where it lived.

There is `bin/bp_genbank2gff3` which is in this repo and part of the [BioPerl](https://metacpan.org/dist/BioPerl) distribution.

There was also a `bin/bp_genbank2gff` which was moved to the [Bio-DB-GFF](https://metacpan.org/dist/Bio-DB-GFF) distributio.

**Comment by cjfields** [comment]
Apologies to @krobison13 about the wait, but all of us 'old-timers' are pretty time constrained these days. 

Coming back around to this, I think we should deprecate Bio::Tools::GFF _particularly if_ there are better options, but we should definitely point in the right direction regardless what we decide.
